src/main/java/com/epam/esm/dao/impl/HibernateGiftCertificateDaoImpl.java

```
@@ -0,0 +1,145 @@
+
+        return typedQuery.getResultList();
+    }
+
+    private <T extends AbstractQuery<GiftCertificate>> void buildQuery(
+            Root<GiftCertificate> root, T query, CertificateRequestBody requestBody) {
+        CriteriaBuilder builder = entityManager.getCriteriaBuilder();
+        Join<GiftCertificate, Tag> tags = root.join("tags");
+        List<Predicate> predicates = new ArrayList<>();
+
+        if (requestBody.getContent() != null) {
+            Predicate predicateForDescription =
+                    builder.like(root.get("name"), "%" + requestBody.getContent() + "%");
+            predicates.add(predicateForDescription);
+            predicateForDescription =
+                    builder.like(root.get("description"), "%" + requestBody.getContent() + "%");
+            predicates.add(predicateForDescription);
+        }
+
+        if (requestBody.getTagNames() != null) {
+            CriteriaBuilder.In<String> inTags = builder.in(tags.get("name"));
+            for (String tagName : requestBody.getTagNames()) {
+                inTags.value(tagName);
+            }
+            predicates.add(inTags);
+            query.groupBy(root.get("id"));
+            query.having(builder.equal(builder.count(root.get("id")), requestBody.getTagNames().size()));
+        }
+        Predicate[] predArray = new Predicate[predicates.size()];
+        predicates.toArray(predArray);
+        query.where(predArray);
+    }
+
+    @Override
+    public int getLastPage(int size) {
+        return persistenceService.getLastPage(GET_CERTIFICATE_COUNT, size);
+    }
+
+    @Override
+    public int addGiftCertificate(GiftCertificate giftCertificate) throws PersistenceException {
+        return persistenceService.addModel(giftCertificate);
+    }
+
+    @Override
+    public boolean deleteGiftCertificate(int id) {
+        try {
+            persistenceService.deleteModel(id);
+            return true;
+        } catch (NoResultException | IllegalArgumentException e) {
```

Why not to catch this exceptions in persistence service?

---

Refactored delete methods to catch all exceptions in service layer